crates/curp/src/server/cmd_worker/mod.rs

```diff
@@ -1,269 +1,244 @@
-                    return;
-                };
-                handle_task(task, &done_tx, ce.as_ref(), curp.as_ref()).await;
-            }
-            _ = shutdown_listener.wait() => break,
+    let pool_entry = match entry.entry_data {
+        EntryData::Command(ref c) => PoolEntry::new(entry.propose_id, Arc::clone(c)),
+        EntryData::ConfChange(ref c) => PoolEntry::new(entry.propose_id, c.clone()),
+        EntryData::Empty | EntryData::Shutdown | EntryData::SetNodeState(_, _, _) => {
+            unreachable!()
         }
-    }
-    while let Ok(task) = dispatch_rx.try_recv() {
-        handle_task(task, &done_tx, ce.as_ref(), curp.as_ref()).await;
-    }
-    debug!("cmd worker exits");
-}
-
-/// Handle task
-async fn handle_task<C: Command, CE: CommandExecutor<C>, RC: RoleChange>(
-    mut task: Task<C>,
-    done_tx: &flume::Sender<(Task<C>, bool)>,
-    ce: &CE,
-    curp: &RawCurp<C, RC>,
-) {
-    let succeeded = match task.take() {
-        TaskType::SpecExe(entry, pre_err) => worker_exe(entry, pre_err, ce, curp).await,
-        TaskType::AS(entry, prepare) => worker_as(entry, prepare, ce, curp).await,
-        TaskType::Reset(snapshot, finish_tx) => worker_reset(snapshot, finish_tx, ce, curp).await,
-        TaskType::Snapshot(meta, tx) => worker_snapshot(meta, tx, ce, curp).await,
     };
-    if let Err(e) = done_tx.send((task, succeeded)) {
-        if !curp.is_shutdown() {
-            error!("can't mark a task done, the channel could be closed, {e}");
-        }
-    }
+    sp.remove(pool_entry.clone());
+    ucp.remove(pool_entry);
 }
 
 /// Cmd worker execute handler
-async fn worker_exe<C: Command, CE: CommandExecutor<C>, RC: RoleChange>(
-    entry: Arc<LogEntry<C>>,
-    pre_err: Option<C::Error>,
+pub(super) fn execute<C: Command, CE: CommandExecutor<C>, RC: RoleChange>(
+    entry: &LogEntry<C>,
     ce: &CE,
     curp: &RawCurp<C, RC>,
-) -> bool {
-    let (cb, sp, ucp) = (curp.cmd_board(), curp.spec_pool(), curp.uncommitted_pool());
+) -> Result<<C as Command>::ER, <C as Command>::Error> {
+    let (sp, ucp) = (curp.spec_pool(), curp.uncommitted_pool());
     let id = curp.id();
-    let success = match entry.entry_data {
+    match entry.entry_data {
         EntryData::Command(ref cmd) => {
-            let er = if let Some(err_msg) = pre_err {
-                Err(err_msg)
-            } else {
-                ce.execute(cmd).await
-            };
-            let er_ok = er.is_ok();
-            cb.write().insert_er(entry.propose_id, er);
-            if !er_ok {
-                sp.lock()
-                    .remove(PoolEntry::new(entry.propose_id, Arc::clone(cmd)));
-                if curp.is_leader() {
-                    ucp.lock()
-                        .remove(PoolEntry::new(entry.propose_id, Arc::clone(cmd)));
-                }
+            let er = ce.execute(cmd);
+            if er.is_err() {
+                remove_from_sp_ucp(&mut sp.lock(), &mut ucp.lock(), entry);
```

Comment:
Please make sure you always keep the same order of `sp` and `ucp`  lock across the whole project.

Reply:
Currently the lock order are consistent. I added a comment in `Context` to specifying the order for acquiring these locks.